test/request_reply_message_receiver_test.go

```diff
@@ -34,81 +40,992 @@ const subscriptionString = "some-subscription"
+						Eventually(receivedMsgChannel, 25*time.Second).Should(HaveLen(sentMessage)) // replies should be sent back
+						// message in the channel should be a request message
+						receivedMessage := <-receivedMsgChannel
+						content, ok := receivedMessage.GetPayloadAsString()
+						Expect(ok).To(BeTrue())
+						Expect(content).To(Equal("Ping"))
+
+						select {
+						// message in the reply channel should be a reply message
+						case replyMessage := <-publishReplyChan:
+							Expect(replyMessage).ToNot(BeNil())
+							content, ok := replyMessage.GetPayloadAsString()
+							Expect(ok).To(BeTrue())
+							Expect(content).To(Equal("Pong"))
+						case <-time.After(10 * time.Second):
+							Fail("Timed out waiting to receive reply message")
+						}
+
+						// check that the message & reply was sent via semp
+						client := helpers.GetClient(messagingService)
+						Expect(client.DataTxMsgCount).To(Equal(int64(sentMessage * 2)))                                                // requests & replies
+						Expect(messagingService.Metrics().GetValue(metrics.DirectMessagesReceived)).To(Equal(uint64(sentMessage * 2))) // send  & reply
+					})
+				}
+
+				It("should properly handle direct massages published to request-reply topic with the Receive() function", func() {
+					helpers.PublishOneMessage(messagingService, topicString, "Ping") // publish direct message
+					// block and wait for direct message
+					msg, replier, err := receiver.ReceiveMessage(-1)
+					Expect(err).ToNot(HaveOccurred()) // not expecting an error
+					Expect(msg).ToNot(BeNil())        // we should receive a message
+					Expect(replier).To(BeNil())       // but no replier since this should be a direct message
+
+					// publish the request messages on another thread to same topic
+					helpers.PublishNRequestReplyMessages(messagingService, topicString, publishTimeOut, 1, "Ping")
+
+					// block and wait for request message
+					msg, replier, err = receiver.ReceiveMessage(-1)
+					Expect(err).ToNot(HaveOccurred())
+					Expect(msg).ToNot(BeNil())
+					Expect(replier).ToNot(BeNil())
+					err = replier.Reply(helpers.NewMessage(messagingService, "Pong")) // send reply back
+					Expect(err).ToNot(HaveOccurred())
+
+					// check that the message & reply was sent via semp
+					client := helpers.GetClient(messagingService)
+					Expect(client.DataTxMsgCount).To(Equal(int64(3)))                                                // 1 direct, 1 request and 1 reply message                                            // requests & replies
+					Expect(messagingService.Metrics().GetValue(metrics.DirectMessagesReceived)).To(Equal(uint64(3))) // 3 messages
+				})
+
+				It("should properly handle direct massages published to request-reply topic with the ReceiveAsync() function", func() {
+					rRMessagesCount := uint64(0)
+					directMessagesCount := uint64(0)
+
+					// apparently the test degrades as the number of published messages increases
```

Comment:
This is due to the receiver discarding received messages when the number of buffered messages exceeds 50 messages. The feature Branch SOL-62455 has added receiver back pressure configuration method on builder to provide easy access to configure option for the receiver buffering and drop behaviour. Note this was configurable with the receiver sting properties before as well.

This test publishes over 50 messages so its possible to have discard issues here. Either change the number of published messages to 50 the default receiver buffer size or configure the receiver with a buffer size of 100 or greater or have your test account for the discarded messages via discard indication on the received messages.

Reply:
Okay, thanks for the explanation. I will update the test case and confirm.